### DISCUSSION TITLE
creating a windows executable using Ifcopenshell

### DISCUSSION THREAD

**Discussion by CMbeebop** [IfcOpenShell/IfcOpenShell#2798] [Q&A] [answered]
Hi,

I am using PyInstaller to create a windows executable for my application, which uses ifcopenshell for reading an ifc file and iterate on the elements collecting their geometrical data and psets. It seems the executable has problems with the ifcopenshell.util module (I'm getting this error report) 

> Traceback (most recent call last):
>   File "myScript.py", line 261, in <module>
>   File "ifcopenshell\util\element.py", line 54, in get_psets
>   File "ifcopenshell\util\element.py", line 173, in get_type
>   File "ifcopenshell\entity_instance.py", line 140, in __getattr__
>   File "importlib\__init__.py", line 126, in import_module
>   File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1128, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
>   File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1128, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
>   File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1142, in _find_and_load_unlocked
> ModuleNotFoundError: No module named 'ifcopenshell.express'
> [9108] Failed to execute script 'myScript' due to unhandled exception!

Although the script works fine in python. By commenting the code using ifcopenshell.util, I have checked that other modules as ifcopenshell.geom are working fine in the executable.

Do you have any idea about what could be going on?
is there any guide or additional information for creating windows executables using ifcopenshell?

regards, Carlos

**Comment by aothms** [comment, marked as answer]
> Do you have any idea about what could be going on?

Yes. We have a dynamic import there https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/entity_instance.py#L157

So PyInstaller cannot statically (without running code) determine which modules need to be included.

You'll have to manually specify those modules in your pyinstaller config.

**Reply by CMbeebop** [reply]
Thanks for your reply,

ok, I have seen pyInstaller has plenty of options that could be relevant https://pyinstaller.org/en/stable/usage.html
I tried using --hiddenimport on ifcopenshell.express as follows 

py -m PyInstaller --clean --icon=myIcon.ico --hiddenimport "ifcopenshell.express" myScript.py

but still not working.
Could you share an example of pyinstaller configuration command or a .spec file for using Ifcopenshell?

**Comment by aothms** [comment]
Did you also try adding ifcopenshell.express.rules.IFC2X3 and every schema
you intend to support?

Sent from a mobile device. Excuse my brevity. Kind regards, Thomas

Op do 23 feb. 2023 08:18 schreef CMbeebop ***@***.***>:

**Reply by CMbeebop** [reply]
like this?
py -m PyInstaller --hiddenimport "ifcopenshell.express.rules.IFC2X3" myScript.py

i'm getting an error running the executable on a ifc file which should be on IFC2X3 format.

> back (most recent call last):
>   File "MyScript.py", line 260, in <module>
>   File "ifcopenshell\util\element.py", line 54, in get_psets
>   File "ifcopenshell\util\element.py", line 173, in get_type
>   File "ifcopenshell\entity_instance.py", line 140, in __getattr__
>   File "importlib\__init__.py", line 126, in import_module
>   File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1128, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
>   File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1128, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
>   File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
>   File "PyInstaller\loader\pyimod02_importers.py", line 352, in exec_module
>   File "ifcopenshell\express\__init__.py", line 29, in <module>
> FileNotFoundError: [Errno 2] No such file or directory: 'C:\\ifc\\dist\\MyScript\\ifcopenshell\\express\\express_parser.py'
> [2708] Failed to execute script 'MyScript' due to unhandled exception!

**Comment by Martin15135215** [comment]
> is there any guide or additional information for creating windows executables using ifcopenshell?

i use this my_script.spec file:

It just use the datas variables to copy the whole directory where ifcopenshell and ifcpatch lives, it's a bit heavy, but it's get the job done 

```python
datas = [	(r"C:\Users\Martin\miniforge\envs\ifcopenshell\Lib\site-packages\ifcopenshell\\", "ifcopenshell"), 
			(r"C:\Users\Martin\miniforge\envs\ifcopenshell\Lib\site-packages\ifcpatch\\", "ifcpatch")]
    
block_cipher = None

a = Analysis(
    ['my_script.py'],
  pathex=[],
    binaries=[],
    datas=datas,
    hiddenimports=[],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=['statmodels', 'setuptools', 'scipy', 'pyinstaller', 'pyinstaller-hooks-contrib', 'python-dateutil','tkinter'],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='my_script',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
coll = COLLECT(
    exe,
    a.binaries,
    a.zipfiles,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='my_script',
)

```

**Comment by CMbeebop** [comment]
that actually worked out, thank you :)

just used the --add-data option of pyinstaller to link ifcopenshell to my folder C:\Users\Me\....\ifcopenshell\Lib\site-packages\ifcopenshell, probably it was looking somewhere else with out the add-data.
